Refine grouping condition for single assertion conflict

sonalmahajan15 opened the pull request:
This PR further refines the grouping condition updated in PR #229 to add function name to the key. This is particularly useful when the text of the producer and consumer messages is not sufficient to distinguish the conflicts.

yuxincs commented on diagnostic/conflict.go:
```diff
@@ -72,6 +77,36 @@ func groupConflicts(allConflicts []conflict) []conflict {
 			key = p.producerRepr + ";" + p.consumerRepr
 			if p.producerPosition.IsValid() {
 				key = p.producerPosition.String() + ": " + p.producerRepr
+			} else {
```
Can we add some comments here listing an example where we would hit this case for future readers? :)

sonalmahajan15 replied:
Done :)

yuxincs commented on diagnostic/conflict.go:
```diff
@@ -72,6 +77,36 @@ func groupConflicts(allConflicts []conflict) []conflict {
+					}
+
+					// Find the enclosing function name for the conflict position
+					ast.Inspect(file, func(n ast.Node) bool {
+						if fd, ok := n.(*ast.FuncDecl); ok {
+							// Check if the conflict position falls within the function's position range. If so, update
+							// the key to include the function name, and end the traversal.
+							functionStart := pass.Fset.Position(fd.Pos()).Offset
+							functionEnd := pass.Fset.Position(fd.End()).Offset
+							if c.position.Offset >= functionStart && c.position.Offset <= functionEnd {
+								key = fd.Name.Name + ":" + key
+								return false
+							}
+						}
+						return true
+					})
+				}
```
I think this AST search can be simplified to just a single pass over `ast.File` node since function declarations are top-level decls and can only appear under `ast.File` node: https://go.dev/ref/spec#Declarations_and_scope

It would save some further traversals :)

sonalmahajan15 replied:
Great idea! Updated the code accordingly.

yuxincs reviewed (approved):
Overall looks good to me (we can also refine this logic to perhaps include a full AST path from root to the diagnostic to further distinguish the diagnostics, but the function name heuristics work for majority of the cases).

I only have one request for some comments and an small optimization :)

sonalmahajan15 commented:
> Overall looks good to me (we can also refine this logic to perhaps include a full AST path from root to the diagnostic to further distinguish the diagnostics, but the function name heuristics work for majority of the cases).
> 
> I only have one request for some comments and an small optimization :)

Agreed, we can decide to refine the logic further to include the full path if the need arises in the future.

Addressed your comments :)
